diff --git a/Build/CMakeLists.txt b/Build/CMakeLists.txt
index 9b1f35d..03d7f70 100644
--- a/Build/CMakeLists.txt
+++ b/Build/CMakeLists.txt
@@ -280,7 +280,7 @@ endif()
 if (ENABLE_INSTALL)
 	install(TARGETS Jolt
 		EXPORT JoltExport
-		DESTINATION lib)
+		DESTINATION bin)
 	foreach(SRC_FILE ${JOLT_PHYSICS_SRC_FILES})
 		string(REPLACE ${PHYSICS_REPO_ROOT} "" RELATIVE_SRC_FILE ${SRC_FILE})
 		get_filename_component(DESTINATION_PATH ${RELATIVE_SRC_FILE} DIRECTORY)
